drafts/interesting-bugs-caught-by-no-constant-binary-expression.md

```diff
@@ -0,0 +1,130 @@
+
+```javascript
+hasData = hasData || data !== {};
+```
+
+_From [Firefox](https://hg.mozilla.org/try/rev/0fe5678fb8b71f4eb26f0a153c52d0be45fc5ac1#l3.34) (also: [Firefox](https://hg.mozilla.org/try/rev/0fe5678fb8b71f4eb26f0a153c52d0be45fc5ac1#l1.13))_
+
+### Expecting Empty Objects to be false/null
+
+Another common error class is expecting empty objects to be nullish or falsy. This is likely an easy mistake to make for folks coming from a language like Python where empty lists and dictionaries are falsy.
+
+```javascript
+const newConfigValue = { ...configProfiles } ?? {};
+```
+
+_From [VS Code](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/workbench/contrib/terminal/browser/terminalProfileQuickpick.ts#L126:L126) (also: VS Code [1](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/platform/terminal/node/ptyService.ts#L369:L369), [2](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts#L484:L484))_
+
+### Is it `>=` or `=>`?
+
+I’ve only seen this particular typo once, but I wanted to include it because think it’s a great example of the unexpected types of bugs this approach can catch.
+
+Here the developer meant to test if a value was greater than or equal to zero (`>= 0`), but accidentally reversed the order of the characters and created an arrow function that returned `0 && startWidth <= 1`!
+
+```javascript
+assert((startWidth) => 0 && startWidth <= 1);
+```
+
+_From [Mozilla](https://phabricator.services.mozilla.com/rMOZILLACENTRAL925b8d1ad45f80faee052492b3b43f5120052405)_
+
+### And More
+
+The above five categories of errors are not exhaustive. When I originally ran the first version of this rule on our (very) large monorepo at work, it found over 500 issues. While many fell into the categories outlines above, there was also a long-tail of other interesting bugs. Some highlights include:
+
+- Thinking `||` allows for set operations: `states.includes('VALID' || 'IN_PROGRESS')`
+- Thinking primitive functions pass through nulls: `Number(x) == null`
+- Not knowing primitive [constructors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/Number) return boxed primitives: `new Number(x) === 10`
+
+I never would have set out to lint for these specific issues individually, but by simply trying to identify anything “useless” we were able to find and correct them.
+
+## Conclusion
+
+As you've now seen no-constant-binary-expression is capable of detecting a variety of different types of bugs. Not because it's programed to look for those specific issues, but because all those bugs have one thing in common: they manifest as useless code. Because developers generally don't intend to write useless code, detecting useless code generally results in detecting bugs.
```

Comment:
```suggestion
As you've now seen, `no-constant-binary-expression` is capable of detecting a variety of different types of bugs. The rule accomplishes this not because it's programmed to look for those specific issues, but because all those bugs have one thing in common: they manifest as useless code. Because developers generally don't intend to write useless code, detecting useless code generally results in detecting bugs.
```